src/overlay/Peer.cpp

```diff
@@ -630,38 +710,256 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
+           Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL);
+    assert(mRemotePeerWantsFlowControl && *mRemotePeerWantsFlowControl);
+    assert(msg.sendMoreMessage().numMessages > 0);
+
+    CLOG_TRACE(Overlay, "Peer {} sent SEND_MORE {}",
+               mApp.getConfig().toShortString(getPeerID()),
+               msg.sendMoreMessage().numMessages);
+
+    // SEND_MORE means we can free some capacity, and dump the next batch of
+    // messages onto the writing queue
+    mOutboundCapacity += msg.sendMoreMessage().numMessages;
+    maybeShedLoadAndSendNextBatch();
+}
+
+bool
+Peer::hasReadingCapacity() const
+{
+    bool enabledAndSupportedByRemotePeer =
+        (mRemoteOverlayVersion >= Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL &&
+         mApp.getConfig().ENABLE_OVERLAY_FLOW_CONTROL);
+    return !enabledAndSupportedByRemotePeer || mCapacity.mTotalCapacity > 0;
+}
+
+static medida::Timer&
+getOutboundFloodTimer(Application& app, StellarMessage const& msg)
+{
+    std::string type;
+    switch (msg.type())
+    {
+    case SCP_MESSAGE:
+        type = "scp";
+        break;
+    case TRANSACTION:
+        type = "tx";
+        break;
+    default:
+        type = "ctrl";
+    }
+
+    return app.getMetrics().NewTimer({"overlay", "outbound-queue", type});
+}
+
+void
+Peer::maybeTrimQueue()
+{
+    auto now = mApp.getClock().now();
+    auto shouldTrim = [&](QueuedOutboundMessage const& msg) -> bool {
+        if (msg.mMessage.type() == TRANSACTION)
+        {
+            // Trim transactions that sit in the queue for too long
+            return (now - msg.mTimeEmplaced) >
+                   Peer::OUTBOUND_TRANSACTION_DROP_TIMEOUT;
+        }
+        else if (msg.mMessage.type() == SCP_MESSAGE)
+        {
+            auto slotIndex = msg.mMessage.envelope().statement.slotIndex;
+            auto minSlotToRemember =
+                mApp.getHerder().getMinLedgerSeqToRemember();
+            if (slotIndex >= minSlotToRemember &&
+                mApp.getHerder().isLatestSCPMessage(msg.mMessage))
+            {
+                return false;
+            }
+            // Trim an SCP message if it's outside of range or we have a newer
+            // statement
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    };
+
+    int dropped = 0;
+    auto it = mFlowControlledQueue.cbegin();
+    while (it != mFlowControlledQueue.cend())
+    {
+        if (shouldTrim(*it))
+        {
+            getOutboundFloodTimer(mApp, it->mMessage)
+                .Update(now - it->mTimeEmplaced);
+            it = mFlowControlledQueue.erase(it);
+            dropped++;
+        }
+        else
+        {
+            ++it;
+        }
+    }
+    if (dropped)
+    {
+        CLOG_TRACE(Overlay, "Dropped {} messages to peer {}", dropped,
+                   mApp.getConfig().toShortString(getPeerID()));
+    }
+}
+
+void
+Peer::maybeShedLoadAndSendNextBatch()
+{
+    releaseAssert(mRemoteOverlayVersion >=
```

Comment:
this function should be reworked a bit:
it's called for every message added and every time we add capacity. This ends up using a lot of resources.

Maybe rate limit how often you trim?